source/MaterialXGenShader/Util.cpp

```diff
@@ -13,119 +13,126 @@ namespace MaterialX
-    {
-        return v < EPS_ZERO;
-    }
-
-    bool isZero(const Color3& c)
+    const float ONE_VALUE = 1.0f;
+    const float ZERO_VALUE = 0.0f;
+    
+    bool isEqual(const float& v, const float& v2)
     {
-        return isZero(c[0]) && isZero(c[1]) && isZero(c[2]);
+        return (v - v2) < EPS_ZERO;
     }
 
-    bool isZero(ValuePtr value)
+    bool isEqual(ValuePtr value, const float& value2)
     {
-        if (value->isA<float>() && isZero(value->asA<float>()))
+        if (value->isA<float>() && isEqual(value->asA<float>(), value2))
         {
             return true;
         }
-        if (value->isA<Color3>() && isZero(value->asA<Color3>()))
+        else if (value->isA<Color3>())
         {
-            return true;
+            const Color3& value3 = value->asA<Color3>();
+            if (isEqual(value3[0], value2) && isEqual(value3[1], value2) && isEqual(value3[2], value2))
+            {
+                return true;
+            }
         }
         return false;
     }
 
-    bool isOne(float v)
-    {
-        return v > EPS_ONE;
-    }
+    using OpaqueTestPair = std::pair<string, float>;
+    using OpaqueTestPairList = vector<OpaqueTestPair>;
 
-    bool isOne(const Color3& c)
+    InputPtr getInputInterface(const string& interfaceName , NodePtr node)
     {
-        return isOne(c[0]) && isOne(c[1]) && isOne(c[2]);
+        InputPtr interfaceInput = nullptr;
+        ElementPtr parent = node->getParent();
+        NodeGraphPtr nodeGraph = parent ? parent->asA<NodeGraph>() : nullptr;
+        if (nodeGraph)
+        {
+            interfaceInput = nodeGraph->getInput(interfaceName);
+            if (!interfaceInput)
+            {
+                NodeDefPtr nodeDef = nodeGraph->getNodeDef();
+                if (nodeDef)
+                {
+                    interfaceInput = nodeDef->getInput(interfaceName);
+                }
+            }
+        }
+        return interfaceInput;
     }
 
-    bool isOne(ValuePtr value)
+    bool hasTransparentInputs(const OpaqueTestPairList& opaqueInputList, NodePtr node)
     {
-        if (value->isA<float>() && isOne(value->asA<float>()))
-        {
-            return true;
-        }
-        if (value->isA<Color3>() && isOne(value->asA<Color3>()))
+        for (auto opaqueInput : opaqueInputList)
         {
-            return true;
+            InputPtr interfaceInput = node->getInput(opaqueInput.first);
+            if (interfaceInput)
+            {
+                if (interfaceInput->getConnectedNode())
+                {
+                    return true;
+                }
+                ValuePtr value = interfaceInput->getValue();
+                if (value && !isEqual(value, opaqueInput.second))
+                {
+                    return true;
+                }
+            }
         }
         return false;
     }
 
-    bool isTransparentShaderNode(NodePtr node)
+    bool isTransparentShaderNode(NodePtr node, OpaqueTestPairList& interfaceNames)
```

Comment:
Curious why this deferred evaluation of interface inputs are needed? Passing this list in and out of all methods makes the code hard to read and I'm not sure why it's needed. Can't the evaluation be done directly instead of passing the information back out and evaluating it somewhere else?

Reply:
Probably easiest to describe with an example:
1. We find a node input with "opacity". This has an "interfacename" association so it cannot be immediately evaluated for all cases. 
2. The immediate check is to see if it's on interface input "is transparent" (connected or not 1) in the parent graph or associated nodedef.
3. When passed back this allows any node which has a functional nodegraph with the node input in 1 to be tested but with the interface name.
e.g. Functional graph:
```
<nodegraph name="NG_test" nodedef="ND_test">
        <output name="out" type="surfaceshader" nodename="standard_surface0" />
    <standard_surface name="standard_surface0" type="surfaceshader" version="1.0.1">
      <input name="opacity" type="color3" value="1, 1, 1" interfacename="surface_cutout" />
    </standard_surface>
  </nodegraph>
  
  <nodedef name="ND_test" node="adsk_test" version="1.0.1" isdefaultversion="true">
    <input name="surface_cutout" type="color3" value="1, 1, 1" />
  </nodedef>
  
  <adsk_test name="shader_transp_test" type="surfaceshader">
    <input name="surface_cutout" type="color3" output="surface_cutout_map_output" nodegraph="inputGraph" />
  </adsk_opaque>  
  <nodegraph name="inputGraph">    
    <image name="surface_cutout_texture" type="color3">
      <input name="file" type="filename" value="blah.png" colorspace="lin_rec709" />
    </image>
    <output name="surface_cutout_map_output" type="color3" nodename="surface_cutout_texture" />
  </nodegraph>
  ```
  Here:
  * std surface opacity is exposed as `surface_cutout`
  *  `shader_transp_test` uses the definition and specifes `surface_cutout` is connected. Which based on heuristic assumes this is transparent. If it was a non-zero value again would be marked as transparent.
  * We don't know to test `surface_cutout` as an interface to test hence we rely on inspection of the definition to indicate it's a canidate to check on `shader_transp_test`.

Reply:
Thanks for the explanation Bernard. I understand now why these external connections can't be evaluated directly.

But would it be possible to pass in the "parent" node instance instead of passing out the found interfacenames? So in this example passing in shader_transp_test, so when a connection with "interfacename" is found in the functional graph you could immediately evaluate the corresponding input on the node? If it's possible the pros would be that you could then early out directly if transparency is found, and you avoid the creation and insertion into the dynamic vectors.

Reply:
I tried out the alternative to a certain extend, and as it may not be an immediate child then it ends up "pushing" down a set of nodes through calls vs returning a list of inputs. I wasn't able to make this node "push" work currently so could be a refinement for later.

Reply:
Sounds good, and we can always refine later if needed.

Reply:
As you can see I figured out the implementation to your logic @niklasharrysson. Thanks for the suggestion as it fixed a few remaining issues :).